### ISSUE TITLE
ECS Agent disconnected and instance become unavailable

### ISSUE THREAD

**Issue by luckyshroom** [aws/amazon-ecs-agent#1960]
**Summary**

After some period of time (day or two or six) instance became unavailable (agent connected = false) only reboot can recover it. I'm trying to find the cause of the problem, but so far unsuccessfully. Cloudwatch logs are clear and don't contains any errors. May be problem in huzelcast cluster networking & AWSJoiner...
In general, I break my head ... I hope you will help to figure it out. Thanks.

**Environment Details**

- AMI: amzn-ami-2018.03.o-amazon-ecs-optimized
- Region: us-east-1b

**Additioal information**

In advance I provide collector-logs from instance: https://drive.google.com/open?id=1s3et1NAeZ07hCVasbcKhyfQmpvbAw_8a

**Comment by adnxn** [comment]
@luckyshroom: to dig into this more - 

for this specific instance and the logs you provide it seems like the the agent was connected steadily.

```
$ grep -rn "Connected to ACS endpoint" . | sort
./ecs-agent.log.2019-03-24-11:39:2019-03-24T11:29:53Z [INFO] Connected to ACS endpoint
...
./ecs-agent.log.2019-03-25-03:78:2019-03-25T03:33:09Z [INFO] Connected to ACS endpoint
```

the `Connected to ACS endpoint` indicates that the agent connect should be `true`.  do you mind describing the failure you're seeing with some more details?


> After some period of time (day or two or six) instance became unavailable

from the logs you provided, it is for a time window of a day and it appears the agent was connected to the backend for that time period. 

if you have some more information regarding timelines of connectivity or agent or task placement behavior - it would be helpful in investigating this further.

**Comment by luckyshroom** [comment]
Yes, I launched a new instance for the test 2 days ago. But somewhere after 24 hours, he went into a disconnected state in ecs (agent = false), and in order to connect to him and collect logs, i was need to reload it. The information I transmitted about (issue may occur after day, 2 or 6) was based on previous experience (I decided to launch a new instance and configure it again, but the result remained the same)

**In summary:** 

Task was placed around 2019-03-24-22 and on the next day 2019-03-25-03 agent connection status was false - I was need to reboot it.

**About AMI2:**

Yeah maybe I need to try to use this one.

**Comment by FernandoMiguel** [comment]
are you sure you arent running out of resources?
RAM , CPU credits? IO credits?

**Comment by luckyshroom** [comment]
@FernandoMiguel : Yeah I was think about it, but metrics says that it's ok (memory consumption peak is about 68%). 
Moreover, the problem usually arose during the period when the instance was in an idle state.

**Comment by luckyshroom** [comment]
Good day for all. This issue became again today between 2019-03-26-17 and 2019-03-27-02. 
I collect new logs: https://drive.google.com/open?id=1s3et1NAeZ07hCVasbcKhyfQmpvbAw_8a
[https://ibb.co/Lnq3Yj4](url)
Thanks)

**Comment by andreialecu** [comment]
Just wanted to say that we have seen this issue as well once, a week or two ago.

`Agent connected` = `false` - had to reboot the host to get it back. Didn't happen again since then.

**Comment by luckyshroom** [comment]
@andreialecu Yeah, but if it's occurs once - it's ok, but if regular - I want to understand why. To "fix" this I was create alarm in health check to reboot it every time if this issue become. Right now I can't find reason why it happen - sad but true((

**Comment by shivas** [comment]
Ok, we have similar issues all over our ECS clusters, so far there were 3 different cases for these:
1. docker service died/containers keep running ecs-agent container stopped
2. EC2 instance checks fail (i guess it's a hard error, so we replace instance)
3. ECS agent disconnects on itself, and then after some time reconnects again (https://github.com/aws/amazon-ecs-agent/blob/0821fbc762d0a4ee2b180eb6cf8a22a5d2a6e4ce/agent/acs/handler/acs_handler.go#L209) this log message is printed

**Comment by luckyshroom** [comment]
After Agent update to 1.27.0 the problem has gone)

**Comment by dimisjim** [comment]
@luckyshroom experiencing disconnection issues with 1.27.0 when the cluster tries to balance a significant reduction of instance count

**Comment by sellers** [comment]
I also experienced this and it started a month or so ago.   We were using ubuntu Trusty and :latest for the image.  When latest became 1.28 we noticed the agent container would stop, and not restart, and then the instance was orphaned from the cluster.   

I pinned the version of the agent to 1.26.1 and 1.27 and it appears more stable (last 24hrs no agents lost).
